---
title: ""
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

\

# Bayesian tree inference using the fossilised birth-death process in RevBayes

In this exercise we'll estimate a time tree of bears under the fossilised birth-death (FBD) process. We'll use a [morphological matrix](data/bears_morphology.nex) that comprises 62 binary characters for 18 living and fossil bears. We'll also use [age information](data/bears_taxa.tsv) associated with our taxa. Download these files and examine the contents. The exercise is adapted from this [published version](https://hal.inria.fr/PGE/hal-02536394) of the exercise, which contains more background information.

For this exercise create a folder called `FBD_exercise_RevBayes` (or whatever you want). Then create two sub-directories: `data` and `scripts`. Recall that for divergence time estimation we need three model components, so we'll create the following scripts:

  1. `main.Rev` for reading the data, organising calls to other scripts and the MCMC settings
  2. `FBD_dating.Rev` for the time tree model
  3. `Mk+G.Rev` for the substitution model
  4. `clock_strict_morpho.Rev` for the clock model

Edit the scripts in a text editor of your choice.

## Read in the data and define helper variables

Start with your `main.Rev` script.
As in previous RevBayes exercises, we'll start by reading in the data and defining some helper variables.

```{r data, eval=FALSE}
taxa <- readTaxonData("data/bears_taxa.tsv")
morpho <- readDiscreteCharacterData("data/bears_morphology.nex")
```

```{r vars, eval=FALSE}
num_taxa <- morpho.ntaxa() # number of taxa
num_branches <- 2 * num_taxa - 2 # number of branches in an rooted tree
```

Define a set of variables for setting up our MCMC.

```{r vars2, eval=FALSE}
moves    = VectorMoves()
monitors = VectorMonitors()
```

## Time tree model

We will use the FBD process as our tree model, which describes the distribution of most likely trees, given the model parameters (speciation, extinction, fossil sampling) and fossil ages. 
Note that because the fossil ages are data, the FBD model is not strictly a prior.

Go to your `FBD_dating.Rev` script to set this up.

Three of the key parameters of the FBD model are speciation (or birth) rate ($\lambda$), extinction (or death) rate ($\mu$) and fossil sampling rate ($\psi$).
We will assume these rates are constant over time and place exponential priors on each of these.

```{r BD, eval=FALSE}
speciation_rate ~ dnExponential(10)
extinction_rate ~ dnExponential(10)
fossil_samp_rate ~ dnExponential(10)
```

Each parameter is a stochastic variable drawn from an exponential distribution with $\delta$ = 10 and a mean of $1/\delta$ = 0.1.

Next specify the moves on these parameters.

```{r BD_moves, eval=FALSE}
moves.append( mvScale(speciation_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(extinction_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(fossil_samp_rate, lambda = 0.5, tune = true, weight = 3.0) )
```

We might also be interested in diversification and turnover, so let's set up a deterministic variable for each of these.

```{r div, eval=FALSE}
diversification := speciation_rate - extinction_rate
turnover := extinction_rate/speciation_rate
```

Next we'll define the probability of extant species sampling ($\rho$). 
Since we sample all extant bears, weâ€™ll specify this probability as a constant variable = 1.0.

```{r rho, eval=FALSE}
rho <- 1.0
```

Now we'll specify a prior on the origin time parameter. 
We'll use the oldest bear fossil as a minimum (35.0) and a recent estimate for the age of carnivores as a maximum (55.0).

```{r node_date, eval=FALSE}
# prior and move on the origin
origin_time ~ dnUnif(37.0, 55.0)
moves.append( mvSlide(origin_time, weight=1.0) )
```

Finally, we'll specify the FBD model, along with moves on the tree topology and branch lengths.

```{r tree_prior, eval=FALSE}
tree ~ dnFBDP(lambda = speciation_rate, mu = extinction_rate, psi = fossil_samp_rate, rho = rho, origin = origin_time, taxa = taxa)

# moves to search tree space
moves.append( mvFNPR(tree, weight = 15.0) )
moves.append( mvCollapseExpandFossilBranch(tree, origin_time, weight = 6.0) )
moves.append( mvNodeTimeSlideUniform(tree, weight = 40.0) )
moves.append( mvRootTimeSlideUniform(tree, origin_time, weight = 5.0) )
moves.append( mvFossilTimeSlideUniform(tree, origin_time, weight = 5.0) )
```

We can create "clade constraints" to monitor the age of specific clades of interest. Let's do that for the clade containing all living bears.

```{r constraint, eval=FALSE}
clade_extant = clade("Ailuropoda_melanoleuca","Tremarctos_ornatus","Melursus_ursinus",
                    "Ursus_arctos","Ursus_maritimus","Helarctos_malayanus",
                    "Ursus_americanus","Ursus_thibetanus")

age_extant := tmrca(tree, clade_extant)
```

We might also be interested in know how many sampled ancestors are in our posterior trees, so let's set up another deterministic variable to monitor this.

```{r SAs, eval=FALSE}
num_samp_anc := tree.numSampledAncestors()
```

```{r hidden_notes, eval=FALSE, echo=FALSE}
#TODO: clock model for molecular data?
#TODO: Remove the age uncertainty?
#TODO: variable coding? RevBayes does what Lewis says
#TODO: reference the age of carnivores
#TODO: acknowledge the raw data sources
#TODO: tuning for some but not all paras? maybe all slide, scale moves
#TODO: A complete set of scripts for this exercise can be downloaded [here](scripts/exercise06-scripts.zip).
#TODO: FigTree doesn't handle tree with sampled ancestors very well, so we'll use an alternative web based program called [IcyTree](https://icytree.org/).
#TODO: Better description of the origin. This parameter approximates the age of the clade.

#`exclude = ["F"]` here means fossil ages will be excluded from the log file - I've done this because the way RevBayes outputs this information is a bit confusing, so we'll ignore that for the moment.
```

## Strict clock model

For simplicity, we'll use a strict clock model to describe the rate of character change, assuming the rate is constant over time and across all branches in our tree.

Open your `clock_strict_morpho.Rev` script. We don't know the rate of evolution, so as before we'll use an exponential prior. 

```{r strict_clock, eval=FALSE}
branch_rates ~ dnExponential(10.0)

moves.append( mvScale(branch_rates, lambda = 0.5, tune = true, weight = 3.0) )
```

This rate will be used for all branches.

## Substitution model

Next we'll specify the Mk substitution model in the `Mk+G.Rev` file, which describes the probability of transitioning from one character state to another. This model is a generalisation of the Jukes-Cantor model for $k$ states.

First we'll use the `fnJC` function to define a **Q matrix** for 2 states (0, 1), since we only have two character states in our morphological character matrix. Remember that under this model rates of change between states and state frequencies are equal.

```{r Q, eval=FALSE}
Q <- fnJC(2)
```

Then we'll add the following lines to add the Gamma sites model, using an exponential prior with mean = 1 on the shape parameter `alpha` and 4 discrete rate categories.

```{r gamma, eval=FALSE}
alpha_morpho ~ dnExponential( 1.0 )
rates_morpho := fnDiscretizeGamma( alpha_morpho, alpha_morpho, 4 )

moves.append( mvScale(alpha_morpho, lambda=0.1, tune = true, weight=3.0) )
```

Then we'll define a stochastic node representing our character matrix and "clamp" that variable to our morphological data. Note that this time we need to use `type = "Standard"`. I'm not sure why RevBayes uses the term "standard" for morphological data. 

```{r phyloCTMC, eval=FALSE}
seq ~ dnPhyloCTMC(tree = tree, Q = Q, type = "Standard", branchRates = branch_rates, siteRates = rates_morpho)
seq.clamp(morpho)
```

## MCMC settings

Back in `main.Rev` add the following lines to include the tree, clock and substitution models.

```{r source, eval=FALSE}
# tree model
source("scripts/FBD_dating.Rev")
# clock model
source("scripts/clock_strict_morpho.Rev")
# substitution model
source("scripts/Mk+G.Rev")
```

The MCMC set up is very similar to before. First we create the variable `mymodel`.

```{r model, eval=FALSE}
mymodel = model(tree)
```

Then we define a set of monitors to capture the output.

```{r monitors, eval=FALSE}
# parameters printed to file
monitors.append( mnModel(filename = "output/bears_FBD_morpho.log", printgen = 100) )
# trees printed to file
monitors.append( mnFile(filename = "output/bears_FBD_morpho.trees", printgen = 100, tree) )
# parameter values printed to screen during the MCMC
monitors.append( mnScreen(printgen = 100, age_extant, num_samp_anc, origin_time) )
```

Finally, we'll set up the MCMC run using the `mcmc` function, specifying our model, the vector of monitors and the vector of moves and run the chain for 100000 generations.

```{r mcmc, eval=FALSE}
mymcmc = mcmc(mymodel, monitors, moves)
mymcmc.run(generations = 100000, tuningInterval = 1000)
```

Note we've added the argument `tuningInterval = 1000`. This tells the chain how long to spend optimising the MCMC moves. Different datasets behave differently $-$ the above option partly automates the selection of parameters that maximise efficiency of the MCMC.

This might take a while to run. Meanwhile, you could move on to the [Next tasks](#next).

## Evaluating the output

As before, open your `.log` file in Tracer.  
It should look something like this. 

```{r, echo=FALSE, out.width="85%",  out.extra='style="padding:10px"'}
knitr::include_graphics("trace5.png")
```

> Explore the output. Can you identify the different parameters? Has the analysis converged?

> What is the age of the most common recent ancestor of bears?

Next let's generate a summary tree.

```{r mcc, eval=FALSE}
trace = readTreeTrace("output/bears_FBD_morpho.trees")
mccTree(trace, file = "output/bears_FBD_morpho.mcc.tre", positiveBranchLengths = TRUE)
```

Once you have your summary tree, open it in FigTree and play around with the settings.
See if you can get something that looks like this.

```{r, echo=FALSE, out.width="85%",  out.extra='style="padding:10px"'}
knitr::include_graphics("mcc_bears_fbd.png")
```

> Which candiates on your tree do you think might be sampled ancestors?

## Next tasks {#next}

Some extra things to try. **Make sure you rename the output files so you don't overwrite the files you generated above!**

### Add molecular sequence data

Try adding the molecular sequence alignment that complements this dataset. You can download this from [here](data/bears_cytb.nex). 
This is an alignment of the mitochondrial gene cytochrome oxidase B. 

There's just a few extra steps we need to add when we read in the data, at the beginning of `main.Rev`.

First add the line that reads in the sequence data and create a constant variable `cytb`. 

```{r data2, eval=FALSE}
taxa <- readTaxonData("data/bears_taxa.tsv")
morpho <- readDiscreteCharacterData("data/bears_morphology.nex")
cytb <- readDiscreteCharacterData("data/bears_cytb.nex")
```

Next we need to make sure the datasets are cross compatible because we don't have molecular sequence data for all species.

```{r missing, eval=FALSE}
cytb.addMissingTaxa( taxa )
morpho.addMissingTaxa( taxa )
```

That's it! 
Now you can go ahead and set up the GTR model as you did in the [previous exercise](https://boussau.github.io/tutorials/COME_tutorials/8_RevBayesTutorial/).

> What differences do you notice in your output?

### Running the analysis under the FBD model only

Try running the analysis under the prior. Simply add `underPrior = TRUE` to the MCMC run function.

```{r prior, eval=FALSE}
mymcmc.run(generations = 10000, tuningInterval = 1000, underPrior = TRUE)
```

Note this is not strictly "under the prior" because the fossil ages are data but RevBayes doesn't consider the FBD as part of the likelihood it prints out, so you can still choose the regular option for running the analysis under the prior to distinguish the signal coming from the FBD/fossil age data vs. the FBD combined with the sequence/morpho data. 

> Compare you results to the posterior obtained using in the previous steps. What information do we gain from the morphological / molecular data?

### Accounting for fossil age uncertainty

Since we basically never know the exact age of any fossils, we typically need to account for age uncertainty. You can explore the impact of age uncertainty by adding the following commands to your `FBD_dating.Rev` script. You first need to download [a new file](data/bears_taxa_age_uncertainty.tsv) containing the correct minimum and maximum bounds for the fossil ages.

The following loop goes through each fossil in our dataset and assigns a uniform distribution between the minimum and maximum bounds of the known interval of uncertainty.

```{r fossil_age, eval=FALSE}

fossils = tree.getFossils()

for(i in 1:fossils.size())
{
    t[i] := tmrca(tree, clade(fossils[i]))

    a_i = fossils[i].getMinAge()
    b_i = fossils[i].getMaxAge()

    F[i] ~ dnUniform(t[i] - b_i, t[i] - a_i)
    F[i].clamp( 0 )
}
```

> Compare your results to the posterior obtained using in the previous steps. Do the results change?