<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>exercise-02.knit</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="exercise-01.html">RevBayes exercise</a>
</li>
<li>
  <a href="exercise-02.html">BEAST2 exercise</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div
id="bayesian-tree-inference-using-the-fossilised-birth-death-process-in-beast2"
class="section level1">
<h1>Bayesian tree inference using the fossilised birth-death process in
BEAST2</h1>
<p>In this exercise we’ll estimate a time tree of bears under the
fossilised birth-death (FBD) process. We’ll use a <a
href="data/bears_morphology_beast.nex">morphological matrix</a> that
comprises 62 binary characters for 18 living and fossil bears. We’ll
also use age information associated with our taxa, which is also
included in this file. Download these files and examine the contents.
The exercise is adapted from <a
href="https://taming-the-beast.org/tutorials/FBD-tutorial/">this
tutorial</a> written by Tracy Heath. The original contains lots of
useful background information.</p>
<p>For this exercise create a folder called
<code>FBD_exercise_BEAST2</code> (or whatever you want).</p>
<p>BEAST2 uses an <code>xml</code> file as input, containing all the
model specifications, including prior parameters and MCMC settings. You
can create the <code>xml</code> file using the BEAST2 accessory program
BEAUti.</p>
<div id="packages" class="section level2">
<h2>Packages</h2>
<p>BEAST2 is similar to RevBayes in providing flexible and modular
options for analysing phylogenetic data in a Bayesian framework. One
practical difference is the package structure used by BEAST2. There is a
suite of models in the core BEAST2 program, but there are many other
models available via packages that you can install using BEAUti.</p>
<p>For the analysis in this tutorial we need to install the
<code>SA</code> and <code>MM</code> packages, both developed by Sasha
Gavryushkina.</p>
<p>Open BEAUTi go to File &gt; Manage Packages, scroll down, select the
<code>SA</code> and <code>MM</code> packages and hit Install at the
bottom. This should bring up a confirmation message.</p>
<p>Restart BEAUTi and we can begin setting up the analysis.</p>
</div>
<div id="reading-the-data" class="section level2">
<h2>Reading the data</h2>
<p>To read in your data, simply drag your nexus file into the partitions
panel. Or go to File &gt; Import Alignment and navigate to your nexus
file.</p>
<p>This will bring up a box that asks you to clarify what type of data
you have. Select Add Morphological Data.</p>
<p>This will then bring up a box that says, ‘Would you like to condition
on recording variable characters only (Mkv)?’ Ordinarily when we collect
morphological data, unlike when we sequence DNA, we don’t typically
collect non-variable characters. If unaccounted for, this results in a
form of ascertainment bias. We can account for this using the ‘v’ (=
variable coding) correction. But for the moment select ‘no’, as this
correction is more computationally expensive, and BEAST2 and RevBayes
handle this in different ways.</p>
<p>If you successfully imported the data, your window should look like
this.</p>
<p><img src="beauti-01.png" width="80%" style="padding:10px" /></p>
</div>
<div id="tip-dates" class="section level2">
<h2>Tip dates</h2>
<p>Move on to the Tip Dates panel and select <code>Use tip dates</code>.
You might have noticed that the names in the nexus file are appended by
a number. This is the age information associated with each species.
BEAUti can parse this information.</p>
<p>Select auto-configure and then <code>use everything</code> &gt; after
last <code>_</code>. Back in the main window also make sure to select
<code>Before the present</code> from the drop down menu.
<code>Since some time in the past</code> is used more commonly for
viruses or bacteria.</p>
</div>
<div id="the-tripartite-model" class="section level2">
<h2>The tripartite model</h2>
<p>Recall that for divergence time estimation we need three model
components (the substitution, clock and tree models). You’ll find these
in the following BEAUti panels:</p>
<ol style="list-style-type: decimal">
<li>The Site Model panel is for the substitution model</li>
<li>The Clock Model panel is for the clock model</li>
<li>The Priors panel is for the time tree model <strong>and</strong> all
the other priors</li>
</ol>
<p>Important note: there are many more options than the ones you’ll see
going through the panels in this tutorial. More options are available
via other packages or you can directly edit the xml.</p>
</div>
<div id="the-site-model" class="section level2">
<h2>The site model</h2>
<p>Head towards the Site Model panel. Because we’re working with
discrete morphological data, we’re only presented with a limited set of
options.</p>
<p>First select <code>estimate</code> for the Substitution Rate.</p>
<p>We’ll use the Lewis Mk model and we can add the Gamma sites model by
putting the number of desired rate categories into the Gamma Category
Count box – let’s use <code>4</code> categories. We also want to
estimate the Shape parameter associated with this model, so select
<code>estimate</code> for this too. The <code>1.0</code> in this box
will become the starting value.</p>
<p>When you’re done your window should look like this.</p>
<p><img src="beauti-02.png" width="80%" style="padding:10px" /></p>
<p>If you had molecular data or multiple partitions, more options would
appear in this panel.</p>
</div>
<div id="clock-model" class="section level2">
<h2>Clock model</h2>
<p>Move onto the Clock Model panel and simply leave the Strict Clock
option in place.</p>
<p><img src="beauti-02b.png" width="80%" style="padding:10px" /></p>
</div>
<div id="priors" class="section level2">
<h2>Priors</h2>
<p>The Priors panel is where a lot of juicy choices happen. Here you
will be presented with the tree model options, along with prior options
for all the other parameters associated with the substitution and clock
models.</p>
<p>First select the first toggle and select the
<code>Fossilised Birth Death Model</code> from the drop down menu
(you’ll find this at the top of this list). You’ll notice that this adds
a bunch of extra parameters to the list.</p>
<p>First let’s concentrate on the options under the first toggle. We’ll
leave all the default options as they are right now. Under this set up,
we’re parametrising the model using the diversification, turnover and
sampling proportion option. We’re also fixing Rho. You can leave this as
1.0, since we sample all living bears.</p>
<p><img src="beauti-03.png" width="80%" style="padding:10px" /></p>
<p>Now we’ve set up the tree model, let’s move onto set up all the other
priors. These are presented below the tree model toggle in alphabetical
order. Remember that it’s bad practice to stick with the default priors,
unless you have good reason to.</p>
<p>First let’s deal with the prior on the shape parameter used for the
shape parameter in the Gamma sites model. For this we’ll use an
exponential distribution with a mean of = 1.</p>
<p><img src="beauti-04.png" width="80%" style="padding:10px" /></p>
<p>Select the clock rate. The default prior on this is a uniform
distribution between 0 and infinity - this is obviously not a good prior
for this parameter! Let’s use an exponential prior with a mean of
0.1.</p>
<p><img src="beauti-05.png" width="80%" style="padding:10px" /></p>
<p>Next we’ll define all the priors of the FBD model parameters,
starting with the diversification rate. Let’s use a gamma prior for this
with shape and scale parameters <span
class="math inline">\(\alpha\)</span> = 2.0 and <span
class="math inline">\(\beta\)</span> = 0.1, resulting in a mean around
0.1.</p>
<p><img src="beauti-05b.png" width="80%" style="padding:10px" /></p>
<p>Then let’s specify the sampling proportion and turnover parameters.
These both lie between 0 and 1, so the default prior of a uniform
distribution between 0 and 1 is a reasonable choice, in the absence of
other information about these parameter values. Let’s leave these as
they are.</p>
<p><img src="beauti-07.png" width="80%" style="padding:10px" /></p>
<p>Now we’ll specify a prior on the origin time parameter. We’ll use the
oldest bear fossil as a minimum (= 35.0) and a recent estimate for the
age of carnivores as a maximum (= 55.0). We also need to ensure the
starting value for the origin is within the bounds. This is currently
set to 100. Change it to something like 40, so it doesn’t violate the
bounds.</p>
<p><img src="beauti-07.png" width="80%" style="padding:10px" /></p>
<p>Finally, we’ll create a clade constraint to monitor the age of all
living bears. Click <code>+ Add Prior</code> and select MRCA (= most
recent common ancestor) prior. Put the following 8 taxa on the righthand
side: <em>Ailuropoda melanoleuca</em>, <em>Tremarctos ornatus</em>,
<em>Melursus ursinus</em>, <em>Ursus arctos</em>, <em>Ursus
maritimus</em>, <em>Helarctos malayanus</em>, <em>Ursus americanus</em>,
<em>Ursus thibetanus</em>. Give the clade a label, e.g.,
<code>living_bears</code>.</p>
<p><img src="beauti-08.png" width="80%" style="padding:10px" /></p>
<p>Leave the <code>monophyletic</code> box unchecked here because we
don’t want to enforce monophyly of this group, as there might be fossils
also belonging to this clade. Note this is how you define nodes for
specifying node calibrations in BEAUti. After defining a clade, you can
then select a distribution from the drop down menu associated with your
node.</p>
</div>
<div id="mcmc-settings" class="section level2">
<h2>MCMC settings</h2>
<p>Navigate to the MCMC panel to finalise the set up. Our dataset is
small, so let’s go with <code>100000</code> for the Chain Length. Leave
the Store Every and Pre Burnin options as they are.</p>
<p>You can then use the <code>tracelog</code> and <code>treelog</code>
toggles to specify the logging frequency. Let’s change the logging
frequency for both of these to <code>100</code>, so we end up with 10000
posterior samples. You can also change the output file names here. The
trees file will keep the same prefix as the log file if you don’t change
the default <code>$(tree)</code>.</p>
<p><img src="beauti-09.png" width="80%" style="padding:10px" /></p>
</div>
<div id="exporting-the-xml" class="section level2">
<h2>Exporting the xml</h2>
<p>To create the input file for BEAST2 we need to export the
<code>xml</code> file. Simply go to File &gt; Save As and navigate to
where you want to save the file on your computer.</p>
<p>Close BEAUti for now. If you have time, you can come back and set up
more analyses using the molecular data.</p>
</div>
<div id="running-beast2" class="section level2">
<h2>Running BEAST2</h2>
<p>This might take a while run pretty quickly. Meanwhile, you could move
on to the <a href="#next">Next tasks</a>.</p>
</div>
<div id="evaluating-the-output" class="section level2">
<h2>Evaluating the output</h2>
<p>As before, open your <code>.log</code> file in Tracer.<br />
It should look something like this.</p>
<p><img src="trace-beast.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Explore the output. Can you identify the different parameters? Has
the analysis converged?</p>
</blockquote>
<blockquote>
<p>What is the age of the most common recent ancestor of bears?</p>
</blockquote>
</div>
<div id="processing-the-trees-files" class="section level2">
<h2>Processing the trees files</h2>
<p>Next let’s generate a summary tree. BEAST has another accessory
program called TreeAnnotator. Open this up. Add a burnin of 10%. From
the <code>Target tree type</code> menu leave the default Maxmimum clade
credibility (MCC) tree option and from the Node heights menu select
<code>Keep target heights</code> (this is because a posterior with lots
of uncertainty, which is expected when we have a small morphological
matrix, can result in negative branch length is we choose median node
ages). Next select your <code>.trees</code> file as the input and create
a file name for your output (e.g.,
<code>bears_morphology_MCC.tre</code>).</p>
<p><img src="treeAnn.png" width="60%" style="padding:10px" /></p>
<p>Once you have your summary tree open it in FigTree and play around
with the settings. See if you can get something that looks like
this.</p>
<p><img src="mcc_bears_fbd_beast.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Which candiates on your tree do you think might be sampled
ancestors?</p>
</blockquote>
<blockquote>
<p>Play around with the HPD intervals using the Node Labels menu. Do you
notice any a bit odd?</p>
</blockquote>
</div>
<div id="next" class="section level2">
<h2>Next tasks</h2>
<p>Try adding the molecular sequence alignment that complements this
dataset. This is an alignment with 1000 sites of the mitochondrial gene
cytochrome oxidase B for 10 living species.</p>
<p>There’s a bug in the latest version of BEAUti that makes it tricky to
set up different partitions, so we’re going to use a precooked xml for
this purpose. You can download this <a
href="bears_morpho_cytb.xml">here</a> and get it running in BEAST.</p>
<blockquote>
<p>What differences do you notice in your output?</p>
</blockquote>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
